### DISCUSSION TITLE
queryString Array Compacting and Key Customisation

### DISCUSSION THREAD

**Discussion by mdavis1982** [livewire/livewire#2036] [1. Help]
I'm using Livewire to build a pretty simple product browser. I've been adapting the data tables lessons from Surge in order to build some filtering functionality, but I now find myself wanting to do something different and can't find any information about being able to achieve what I want.

At the moment, I have the following property in my component:

```php
public $filters = [
    'search' => '',
    'primaryColour' => [],
];
```

The object is to have a bunch of colour filters on the left, and essentially do an `or` between them. My stub filter picker looks as follows:

![image](https://user-images.githubusercontent.com/199807/99268140-b2ddae80-281c-11eb-801d-b3f74d8fc303.png)

This all works fine and I can get the desired filtering behaviour working like a dream.

My problem is trying to then put these filter values in the query string.

I've added the following to component:

```php
protected $queryString = ['filters'];
```

This works, but it's generating some pretty gross URLs:

```
http://myapp.test/products?search=&primaryColour%5B0%5D=brown&primaryColour%5B1%5D=yellow
```

I'm trying to get my URLs to look different, but I'm not sure if any of this is possible with Livewire yet.

### Encoding of URLs
I think the first thing that would probably make these URLs look nicer is if they were decoded. So, instead of the above, the URL would look like this:

```
http://myapp.test/products?search=&primaryColour[0]=brown&primaryColour[1]=yellow
```

Is this something that is possible?

### `except` Rules for Nested Data
In the URL above, there is an empty `search` parameter. Because this is nested inside my `$filters` array, I can't use the normal `'except' => ''` functionality.

Is it possible to still keep this inside the `$filters` array, but have an `'except'` rule for it?

### Key Customisation
At the moment, the URL has `primaryColour` inside it. I know that I can change this by changing the name of the property, but is it possible to change this without changing the property name?

For example, in my app, the products have a primary colour and then a collection of additional colours. It makes sense for me to refer to `primaryColour` in my app, but it might be good in the URL to expose it as `colour`.

Is this possible? Maybe something like:

```php
protected $queryString = [
    'primaryColour' => ['as' => 'colour']
];
```

It's entirely possible that this is already functionality that exists and I just haven't discovered it yet!

### Array Compacting
Is it possible to compact multiple values in an array to a single query string parameter?

For example, instead of this:

```
http://myapp.test/products?search=&primaryColour[0]=brown&primaryColour[1]=yellow
```

Is it possible to achieve this:

```
http://myapp.test/products?search=&primaryColour=brown,yellow
```

Again, this could be exposed with something like:

```php
protected $queryString = [
    'primaryColour' => ['compact' => ',']
];
```

Again, this might be something that's possible and that I've missed it!

Putting all of this together, my ideal scenario would be able to do something along the lines of:

```php
protected $queryString = [
    'filters.search' => ['except' => ''],
    'filters.primaryColour' => ['as' => 'colour', 'compact' => ','],
];
```

Any help on any of this would be really appreciated. And, of course, if any of this isn't currently possible, I'd love to dive in and try to make some of it possible!

Thanks so much for Livewire. It's really awesome! 😄

**Comment by jochemgruter** [comment]
I support this idea. Two days ago I made issue #2043 for this. Hope this can be integrated.

**Comment by RogerKeliu** [comment]
Hey, 

How is going this issue ?
I'm having exactly the same problem on my application, but I have seen that the related issue is closed and this discussion is not helpful. 

I will appreciate a possible solution for that or other ways to try to implement it.

**Comment by HDVinnie** [comment]
Any update?

**Comment by nullthoughts** [comment]
I know it's been a while since there was any activity on this discussion, but I ended up making a few tweaks to support what I needed with urls and filtering behavior.

For these examples, I'm utilizing a public `$filters` property and using it as a `$queryString` property:
```php
public $filters = [
    'make' => [],
    'model' => [],
    ...
];

public $queryString = ['filters'];
```

### Query strings with fresh Livewire install: 
`?filters%5Bmake%5D%5B0%5D=Tesla&filters%5Bmake%5D%5B1%5D=VW`

### Query strings after adding url-decoding (778735d311c21821b26403b10c7ba876fa30f9db):
`?filters[make][0]=Tesla&filters[make][1]=VW`

This change requires no changes in the component.

### Query strings after adding ability to customize queryParams (f11243b98804b93492b9fe9ab94417ab7600e983) and supporting commas with user configurable encoding type (e44e4548db832ef1c489b826761c88bb915c96a8):
`?make=Tesla,VW`

For this specific example, I've removed the `$queryString` property from the Livewire component.

Commas are supported by switching to RFC1738 in the config:
`'url_encoding_type' => PHP_QUERY_RFC1738,`

I'm flattening the array filter values in an added `queryParams()` method (the nice thing is you could format your data as desired for each component here):
```php
public function queryParams()
{
    return collect($this->filters)
        ->map(function ($value) {
            if (is_array($value)) {
                return implode(',', $value);
            }

            return $value;
        })->reject(function ($value) {
            return empty($value);
        });
}
```

Also in my component, I'm populating my `$filters` property with the request query in `mount()`:
```php
public function mount()
{
    collect(request()->input())->filter(function ($value, $filter) {
        return Arr::has($this->filters, $filter);
    })->reject(function ($value, $filter) {
        return empty($value);
    })->each(function ($value, $filter) {
        if (is_array($this->filters[$filter])) {
            $value = explode(',', $value);
        }

        $this->filters[$filter] = $value;
    });
}
```

One improvement would be to still utilize `$queryString` property (and be able to reference `$this->queryString` in the above method), but ignore it in the `SupportBrowserHistory` class if a `queryParams()` method exists in the component.

Finally, since we are rejecting empty parameters (ex. `make=`) from our queryParams collection, we enable prior parameters to be removed by checking if the component parameter value is empty: ba2cc18fd12cdd3adfa5c833682de29feb32099f (this prevents the query from becoming `?make=&model=&...`

### Before a PR
I could use help testing it to make sure it's fully backwards compatible, you can view the differences at: https://github.com/livewire/livewire/compare/master...nullthoughts:master

I wasn't able to get Dusk up and running quickly, nor have I utilized it previously for package testing. Need to find some time to work on that (any help would be appreciated).

Testing also needs to be done for nested components.

**Reply by joshhanley** [reply]
@nullthoughts to get dusk up and running, have a look at this discussion https://github.com/livewire/livewire/discussions/3494

Can I suggest, if you already done the work, to open a PR, even if it's just a draft while you work on tests, so it can be reviewed?

Please note, you currently have made your changes on your master branch of your fork, but maintainers can't modify that. So you are better off creating a new branch in your fork and then PRing the branch, that way if maintainers need to contribute or make any changes, they can.

Hope this helps!

**Reply by nullthoughts** [reply]
@joshhanley Thank you! The dusk info was very helpful.

As per your suggestions, I've gone ahead and created a new branch and opened a PR: #3524

I also realized that the following would also work (without additional changes to the `SupportBrowserHistory` class):

```php
public $queryString = ['filters'];

public function queryParams()
{
    $params = collect($this->queryString)->map(function ($param) {
        return collect($this->{$param})
            ->map(function ($value) {
                if (is_array($value)) {
                    return implode(',', $value);
                }

                return $value;
            })->reject(function ($value) {
                return empty($value);
            });
    });

    return $params->collapse();
}
```

(might be more useful in certain situations, instead of specifying the property name(s) in `queryParams()`)

**Reply by joshhanley** [reply]
@nullthoughts no worries 🙂 ok great, thanks!

**Reply by keizah7** [reply]
https://laracasts.com/discuss/channels/livewire/livewire-query-string

I'm trying to make it work with `spatie/laravel-query-builder` but after value change component don't see query string in `request()`..

First array is after manual reload,
second after filter value change

`request()->all`

`/apartments?filter[city]=Palanga`

![image](https://user-images.githubusercontent.com/49227202/131491215-019adcbe-7e06-49ee-9142-2f61fee0d759.png)

**Reply by joshhanley** [reply]
@keizah7 I have a feeling that using an array in the query string isn't supported, but I'm not exactly sure.

@nullthoughts are you able to provide some not null thoughts on this? 😬 You're more familiar with the query string stuff than me and with your recent PR and ideas, might have a better clue than me as to whether this will work.

**Reply by keizah7** [reply]
@joshhanley  yes.

```
public $filter = [
    'city' => [],
];

public $queryString = ['filter'];
```

```
<select  class="block w-full p-1" wire:model="filter.city"></select>
```

**Reply by keizah7** [reply]
Just noticed that I can parse `updates` and merge it to request but filter still doesn't work
```
request()->request = new ParameterBag(['filter' => ['city' => 'Palanga']])
```

**Reply by keizah7** [reply]
https://forum.laravel-livewire.com/t/anybody-using-spatie-laravel-query-builder-with-livewire/299/4
solved

**Comment by amiranagram** [comment]
> Key Customisation
> 
> At the moment, the URL has primaryColour inside it. I know that I can change this by changing the name of the property, but is it possible to change this without changing the property name?
> 
> For example, in my app, the products have a primary colour and then a collection of additional colours. It makes sense for me to refer to primaryColour in my app, but it might be good in the URL to expose it as colour.
> 
> Is this possible? Maybe something like:
> 
> ```
> protected $queryString = [
>     'primaryColour' => ['as' => 'colour']
> ];
> ```
> 
> It's entirely possible that this is already functionality that exists and I just haven't discovered it yet!

Are you guys willing to support query string aliases?

**Comment by amiranagram** [comment]
Also when using arrays with `except` filling one item in that array shows all the other, even if empty.

For example:

```
public $filters = [
    'registered-at-min' => '',
    'registered-at-max' => '',
];

protected $queryString  = [
    'filters' => [
        'except' => [
            'registered-at-min' => '',
            'registered-at-max' => '',
        ]
    ],
];
```

Now let's fill only `$filters['registered-at-min']`.

The URL now becomes: `livewire.test/component?filters[registered-at-min]=2021-08-02&filters[registered-at-max]=` with `$filters['registered-at-max']` showing in URL.

Most likely this can only be solved using recursion.

I've hacked a bit on it in my Livewire fork. Doesn't seem that hard to fix.

**P.S. For both this comment and the comment above, I'm willing to PR if you guys want to support it.** @joshhanley

**Reply by nullthoughts** [reply]
@amiranagram I'll have to do some testing with a setup like you provided, but take a look at PR #3679 for granular control over the query string.

Edit: I'm referring to rejecting empty values in dehydratePropertiesToQueryParams() of the PR

**Reply by amiranagram** [reply]
Oh, that seems like a better fix. I was more looking to filter them out in `mergeComponentPropertiesWithExistingQueryParamsFromOtherComponentsAndTheRequest` method in `SupportBrowserHistory` class.

But, are you sure this can be made to work with `except`, since the issue I raised was about `except`, not empty values?

**Reply by nullthoughts** [reply]
I honestly haven't tested it in a setup similar to yours, but would expect $except to omit the empty registered-at-max property. I didn't want to define except for every property in array, so I'm using the dehydrate method to do so. The PR also applies to customizing the keys (your first question).

**Comment by luanfreitasdev** [comment]
> ```
> 'filters.search' => ['except' => ''],
>     'filters.primaryC
> ```

I created a PR for this functionality with dot notation

https://github.com/livewire/livewire/pull/4913

**Comment by theupriser** [comment]
For the compacting, I made something similar to it, but it was not included with livewire sadly. 

https://github.com/livewire/livewire/discussions/4840
https://github.com/livewire/livewire/pull/4838

**Reply by amiranagram** [reply]
Well, Caleb  mentioned that query strings will receive a complete rewrite for v3, I'm guessing that's why.

**Reply by theupriser** [reply]
Ahh okay, cool! But I guess v3 is far away?

**Reply by amiranagram** [reply]
No idea. Cool guitar BTW!

**Comment by Coyote6** [comment]
I am sure everyone has probably figured out their own solution for this by now or are wanting on V3, but here is what I have come up with for the $queryString with an array and wanting to hide the empty strings.  First remove your filter from the $queryString and then on booted() method run this modification of @nullthoughts's code:
```php
protected $queryString = [
	'sorts',
//	'filters'
];

public function booted () {
	collect (request()->input())->filter (function ($value, $filter) {
		if ($filter == 'filters') {	
        		if (is_array ($value)) {
			    	foreach ($value as $k => $v) {
				    	$this->filters[$k] = $v;
			    	}
			}
		}
    	});
}
```
You would need to do this foreach of your queryString values that are arrays but it seems to be working for me. Downside is you lose aliasing or would need to do more modifications to accomplish that.
